ParProcCo/job_scheduler.py

```diff
@@ -0,0 +1,261 @@
+        self.start_time = datetime.now()
+        self.job_history: Dict[int, Dict[int, StatusInfo]] = {}
+        self.job_completion_status: Dict[str, bool] = {}
+
+    def check_queue_list(self, queue: str) -> str:
+        if not queue:
+            raise ValueError(f"queue must be non-empty string")
+        queue = queue.lower()
+        with os.popen('qconf -sql') as q_proc:
+            q_name_list = q_proc.read().split()
+        if queue in q_name_list:
+            return queue
+        else:
+            raise ValueError(f"queue {queue} not in queue list: {q_name_list}\n")
+
+    def check_project_list(self, project: str) -> str:
+        if not project:
+            raise ValueError(f"project must be non-empty string")
+        with os.popen('qconf -sprjl') as prj_proc:
+            prj_name_list = prj_proc.read().split()
+        if project in prj_name_list:
+            return project
+        else:
+            raise ValueError(f"{project} must be in list of project names: {prj_name_list}\n")
+
+    def check_jobscript(self, jobscript: Path) -> Path:
+        if not jobscript.is_file():
+            raise FileNotFoundError(f"{jobscript} does not exist\n")
+
+        if not (os.access(jobscript, os.R_OK) and os.access(jobscript, os.X_OK)):
+            raise PermissionError(f"{jobscript} must be readable and executable by user\n")
+
+        try:
+            js = jobscript.open()
+            js.close()
+        except IOError:
+            logging.error(f"{jobscript} cannot be opened\n")
+            raise
+
+        else:
+            return jobscript
+
+    def get_output_paths(self) -> List[Path]:
+        return self.output_paths
+
+    def get_success(self) -> bool:
+        return all(self.job_completion_status.values())
+
+    def timestamp_ok(self, output: Path) -> bool:
+        mod_time = datetime.fromtimestamp(output.stat().st_mtime)
+        if mod_time > self.start_time:
+            return True
+        return False
+
+    def run(self, jobscript: Path, input_path: Path, slices: List[slice]) -> bool:
+        self.job_history[self.batch_number] = {}
+        self.job_completion_status = {f"{s.start}:{s.stop}:{s.step}": False for s in slices}
+        self._run_and_monitor(jobscript, input_path, slices)
+        return self.get_success()
+
+    def _run_and_monitor(self, jobscript: Path, input_path: Path, slices: List[slice]) -> None:
+        jobscript = self.check_jobscript(jobscript)
+        self.job_details: List[List] = []
+
+        session = JobSession()  # Automatically destroyed when it is out of scope
+        self._run_jobs(session, jobscript, input_path, slices)
+        self._wait_for_jobs(session)
+        self._report_job_info()
+
+    def _run_jobs(self, session: JobSession, jobscript: Path, input_path: Path, slices: List[slice]) -> None:
+        logging.debug(f"Running jobs on cluster for {input_path}")
+        slice_params = [[f"{s.start}:{s.stop}:{s.step}"] for s in slices]
```

Comment:
Keep as slices here and do conversion in `_create_template` as later we may want to make this more flexible and allow scripts to take different arguments and/or in different form